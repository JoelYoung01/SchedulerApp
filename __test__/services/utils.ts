import { Time, Shift, DayOftheWeek, Employee } from "../../entities/types";

export const Monday = DayOftheWeek.Monday;
export const Tuesday = DayOftheWeek.Tuesday;
export const Wednesday = DayOftheWeek.Wednesday;
export const Thursday = DayOftheWeek.Thursday;
export const Friday = DayOftheWeek.Friday;
export const Saturday = DayOftheWeek.Saturday;
export const Sunday = DayOftheWeek.Sunday;

export function shift(
  from: string,
  to: string,
  day: DayOftheWeek,
  owner?: string
): Shift {
  const s = new Shift("", Time.fromString(from), Time.fromString(to), day);
  if (owner !== undefined) {
    s.owner = owner;
  }
  return s;
}
export function allDay(day: DayOftheWeek): Shift {
  return new Shift("", Time.fromString("00:00"), Time.fromString("24:00"), day);
}

export function person(
  name: string,
  minHours: number,
  maxHours: number,
  available: Shift[]
): Employee {
  const ret = new Employee(name, minHours, maxHours);
  ret.available = available;
  return ret;
}
