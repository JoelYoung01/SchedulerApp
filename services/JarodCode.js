class Post {
    constructor(startTime, endTime, day) {
        this.startTime = startTime;
        this.endTime = endTime;
        this.day = day;
        this.owner = "";
    }
}

class Person {
    constructor(name,minHour,maxHour){
        this.name = name;
        this.minHour = minHour;
        this.maxHour = maxHour;
        this.currentHour = 0;
        this.available = [];
        this.busy = [];
    }
}

schedule = [];
schedule.push(new Post(8,12,1));
schedule.push(new Post(15.45,20,1));
schedule.push(new Post(14,16,2));

staff = [];
staff.push(new Person("Alice",1,4));
staff.push(new Person("Bob",1,12));
staff.push(new Person("Clair",1,12));
staff[0].available.push(new Post(0,24,1));
staff[0].available.push(new Post(0,24,2));
staff[1].available.push(new Post(0,24,1));
staff[1].available.push(new Post(0,24,2));
staff[2].available.push(new Post(15,24,2));

isOverlapping = function(shift1,shift2){
    if(shift1.day != shift2.day){
        return shift1.startTime <= shift2.endTime && shift2.startTime <= shift1.endTime;
    }

    return false;
}

shuffle = function(array) {
  let currentIndex = array.length,  randomIndex;
  // While there remain elements to shuffle.
  while (currentIndex != 0) {
    // Pick a remaining element.
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex--;
    // And swap it with the current element.
    [array[currentIndex], array[randomIndex]] = [array[randomIndex],array[currentIndex]];
  }
  return array;
}

getDuration = function(shift){
    //Gets only the minutes
    holder = shift.endTime - shift.startTime;
    //Math Magic
    if(holder - Math.floor(holder) >= .6){
        holder -= .4;
    }

    return holder;
};

generate = function(){
    for(s=0;s<schedule.length;s++){
        shuffle(staff); 
        /**
        * To Do
        * Instead of shuffle(staff), send the Employee that just gained a shift to the back
        * That would also allow you to check if the previous shift is "" and if it is you can end the function early for an error
        */
        foundEmployee: for(e=0;e<staff.length;e++){
            for(a=0;a<staff[e].available.length;a++){
                if(staff[e].available[a].day == schedule[s].day && schedule[s].startTime >= staff[e].available[a].startTime && schedule[s].endTime <= staff[e].available[a].endTime && staff[e].currentHour + getDuration(schedule[s]) <= staff[e].maxHour){
                    //Can't work if they have an overlapping shift
                    overlap = false;
                    for(y=0;y<staff[e].busy.length;y++){
                        if(isOverlapping(schedule[s],staff[e].busy[y])){
                            overlap = true;
                            break;
                        }
                    }
                    //If the employee is available and isn't busy, then they gain the shift
                    if(!overlap){
                        schedule[s].owner = staff[e].name; //Assign shift
                        staff[e].currentHour += getDuration(schedule[s]); //Increase Hours
                        staff[e].busy.push(schedule[s]); //Add the shift to their busy
                        break foundEmployee; //return to the 1st for loop & get a new shift
                    }
                }
            }
        }
    }
}

reset = function(){
    for(p=0;p<staff.length;p++){
        staff[p].busy = [];
        staff[p].currentHour = 0;
    }
    for(p=0;p<schedule.length;p++){
        schedule[p].owner = "";
    }
}

//output
fail = 0;
do{
    valid = true;
    if(fail++ >= 1000){
        fail--;
        console.log("Impossible");
        break; //Return Error
    }
    generate();
    //Makes sure all shifts were assigned
    for(i=0;i<schedule.length;i++){
        if(schedule[i].owner ==  ""){
            valid = false;
            reset();
            break;
        }
    }
    if(valid){
        for(i=0;i<staff.length;i++){
            if(staff[i].currentHour < staff[i].minHour){
                valid = false;
                reset();
                break;
            }
        }
    }
} while(!valid);

console.log("It took " + fail + " attempt(s)");
console.log(schedule);