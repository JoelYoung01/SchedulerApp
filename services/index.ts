export * from "./scheduleState";
export * from "./storageService";
export * from "./util";
export * from "./waveform_collapse";
