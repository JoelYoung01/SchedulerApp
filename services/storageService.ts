import localforage from "localforage";

/*Create a TS Service and interface that can save and retrieve schedule data by a given id or key (up to you, must be unique identifier)

Save data to browser storage, look into localforage npm package or related

include method for getting all local storage schedules

CRUD functionality for local storage with schedules*/

export type Schedule = any;

export class StorageService {
  //getting specific schedule based on key data
  async read(name: string): Promise<Schedule> {
    // The same code, but using ES6 Promises.
    return await localforage.getItem(name);
  }

  //reading the schedule list and seeing if the current schedule is already in the storage
  //if it isnt then add the currently held schedule to the schedule list
  update(name: string, val: Schedule) {
    localforage.setItem(name, val);
  }

  //deleting a specific schedule list and seeing if the specified schedule is even in the storage
  delete(name: string) {
    localforage.removeItem(name);
  }

  async returnAll(): Promise<{ [key: string]: Schedule }> {
    let ret: { [key: string]: Schedule } = {};
    await localforage.iterate((value: Schedule, key: string, _: number) => {
      ret[key] = value;
    });
    return ret;
  }
}
