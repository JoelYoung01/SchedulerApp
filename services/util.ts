export async function sleepTask(ms: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, ms));
}

export function or<T>(a: T | undefined, b: T): T {
  if (a !== undefined) {
    return a;
  } else {
    return b;
  }
}

const DEBUG_LOG = false;

export function log(message?: any, ...more: any[]): void {
  if (DEBUG_LOG) {
    console.log(message, ...more);
  }
}
