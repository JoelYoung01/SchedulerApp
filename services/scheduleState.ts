import { useState } from "react";
import { Color, Employee, Schedule, Shift, Time } from "../entities";
import { generate } from "./waveform_collapse";
import { or, log } from "./util";
import { LocalStorage } from "./storageService";
import { Dayjs } from "dayjs";

interface Add {
  add: Employee | Shift;
}

interface Remove {
  remove: Employee | Shift;
}

interface Set {
  set: Schedule;
}

interface UpdateEmployee {
  update: Employee;
  color?: Color;
  name?: string;
  maxHours?: number;
  minHours?: number;
}

interface UpdateShift {
  update: Shift;
  name?: string;
  start?: Time;
  end?: Time;
}

interface UpdateBase {
  update: "default";
  maxHours?: number;
  minHours?: number;
  name?: string;
  weekDate?: Dayjs;
}

export type ScheduleAction =
  | Add
  | Remove
  | Set
  | UpdateEmployee
  | UpdateShift
  | UpdateBase;

export async function updateSchedule(
  state: Schedule,
  action: ScheduleAction
): Promise<Schedule> {
  const storage = new LocalStorage();
  // Create a deep copy so the generate schedule is allowed to mutate it
  let scheduleCopy = Schedule.createDeepCopy(state);

  let needToRecompute = true;

  if ("set" in action) {
    scheduleCopy = Schedule.createDeepCopy(action.set);
  } else if ("add" in action) {
    if (action.add instanceof Employee) {
      scheduleCopy.addEmployee(action.add);
    } else if (action.add instanceof Shift) {
      scheduleCopy.addShift(action.add);
    }
  } else if ("remove" in action) {
    if (action.remove instanceof Employee) {
      scheduleCopy.removeEmployee(action.remove);
    } else if (action.remove instanceof Shift) {
      scheduleCopy.removeShift(action.remove);
    }
  } else if ("update" in action) {
    if (action.update === "default") {
      needToRecompute = false;
      if (action.maxHours !== undefined) {
        scheduleCopy.maxHoursWorked = action.maxHours;
      }
      if (action.minHours !== undefined) {
        scheduleCopy.minHoursWorked = action.minHours;
      }
      if (action.name !== undefined) {
        // Delete old name
        await storage.delete(scheduleCopy.name);
        scheduleCopy.name = action.name;
        const newUrl = window.location;
        newUrl.hash = `#${scheduleCopy.name}`;
        window.location.replace(newUrl.href);
      }
      if (action.weekDate !== undefined) {
        // Delete old name
        scheduleCopy.weekDate = action.weekDate;
      }
    } else if (action.update instanceof Shift) {
      const a = action as UpdateShift;
      const e = scheduleCopy.removeShift(a.update);
      scheduleCopy.addShift(
        new Shift(or(a.name, e.name), or(a.start, e.start), or(a.end, e.end))
      );
    } else if (action.update instanceof Employee) {
      const a = action as UpdateEmployee;
      const e = scheduleCopy.removeEmployee(a.update);
      scheduleCopy.addEmployee(
        new Employee(
          or(a.name, e.name),
          or(a.minHours, e.min_hours),
          or(a.maxHours, e.max_hours),
          or(a.color, e.color),
          e.available
        )
      );
    }
  }

  await storage.update(scheduleCopy.name, scheduleCopy);

  // Generate the schedule
  if (needToRecompute) {
    const schedulePromise = await generate(
      scheduleCopy.shifts,
      scheduleCopy.employees
    );
    if (schedulePromise !== undefined) {
      // If the scheduler failed, error out
      console.error("Unable to build schedule completely");
    }
    log(scheduleCopy);
  }
  return scheduleCopy;
}

export type Dispatch<A> = (action: A) => void;

/**
 * Async equavalent to React.useReducer
 * */
export function useAsyncReducer<T, A>(
  reducer: (state: T, action: A) => Promise<T>,
  initialState: T
): [T, Dispatch<A>] {
  const [state, setState] = useState(initialState);
  let current: Promise<T> = Promise.resolve(state);

  const dispatch = (action: A): void => {
    // We need to be able to handle multiple in-flight requests
    //
    // The idea is this will force multiple requests to wait in line. Ideally,
    // we would also be able to tell the reducer that there are more (i.e. don't
    // bother fully generating the schedule), but that doesn't seem like something
    // JS can do.
    current = current.then(async (state) => {
      const newState = await reducer(state, action);
      setState(newState);
      return newState;
    });
  };

  return [state, dispatch];
}
