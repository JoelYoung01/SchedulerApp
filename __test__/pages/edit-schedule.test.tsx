import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import EditSchedule from "../../pages/edit-schedule";
import { generate, sleepTask } from "../../services";

jest.mock("../../services/waveform_collapse");

beforeAll(() => {
  jest.mocked(generate).mockResolvedValue(undefined);
});

const toLocaleString = Date.prototype.toLocaleString;
function fixedLocale(
  this: Date,
  locales?: string | string[] | undefined | Intl.LocalesArgument,
  options?: Intl.DateTimeFormatOptions | undefined
): string {
  if (locales === undefined) {
    locales = "en-US";
  }
  if (options === undefined) {
    options = {};
  }
  options.timeZone = "UTC";
  return toLocaleString.call(this, locales, options);
}
// eslint-disable-next-line no-extend-native
Date.prototype.toLocaleString = fixedLocale; // Make dates always return the same value
jest
  .useFakeTimers({
    advanceTimers: true
  })
  .setSystemTime(new Date("2020-01-01"));

test("verify date", () => {
  expect(new Date()).toEqual(new Date("2020-01-01"));
});

test("EditSchedule Renders", () => {
  render(<EditSchedule />);
});

test("Can use Reducer", async () => {
  const user = userEvent.setup();
  const page = render(<EditSchedule />);

  // Add a shift
  await user.click(screen.getByText(/Add Shift/i));

  // Enter shift information
  await user.click(screen.getByLabelText(/Select Start Day/i));
  await user.click(screen.getByText(/Sunday/i));
  await user.type(screen.getByLabelText(/Select Start Time/i), "10:10 PM");
  await user.type(screen.getByLabelText(/Select End Time/i), "02:10 AM");

  // Clicking Submit button
  expect(page.getByText(/Submit/)).not.toBeDisabled();
  await user.click(screen.getByText(/Submit/i));
  expect(page.queryByText(/add a shift/i)).toBe(null);

  await sleepTask(500);

  expect(screen.getByText(/10:10PM/i)).toBeInTheDocument();
  expect(screen.getByText(/2:10AM/i)).toBeInTheDocument();
}, 100000);
