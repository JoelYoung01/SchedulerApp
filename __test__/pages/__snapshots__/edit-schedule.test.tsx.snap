// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSchedule Renders 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 printed css-tpk66z-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1umt6p6-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-dqfxkm-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft css-16taog-MuiTypography-root"
                >
                  12.29
                </h6>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignRight css-1js305m-MuiTypography-root"
                >
                  Sun
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1idn90j-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1umt6p6-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-dqfxkm-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft css-16taog-MuiTypography-root"
                >
                  12.30
                </h6>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignRight css-1js305m-MuiTypography-root"
                >
                  Mon
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1idn90j-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1umt6p6-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-dqfxkm-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft css-16taog-MuiTypography-root"
                >
                  12.31
                </h6>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignRight css-1js305m-MuiTypography-root"
                >
                  Tue
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1idn90j-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1umt6p6-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-dqfxkm-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft css-16taog-MuiTypography-root"
                >
                  1.1
                </h6>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignRight css-1js305m-MuiTypography-root"
                >
                  Wed
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1idn90j-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1umt6p6-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-dqfxkm-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft css-16taog-MuiTypography-root"
                >
                  1.2
                </h6>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignRight css-1js305m-MuiTypography-root"
                >
                  Thu
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1idn90j-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1umt6p6-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-dqfxkm-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft css-16taog-MuiTypography-root"
                >
                  1.3
                </h6>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignRight css-1js305m-MuiTypography-root"
                >
                  Fri
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1idn90j-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1umt6p6-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-dqfxkm-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft css-16taog-MuiTypography-root"
                >
                  1.4
                </h6>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 MuiTypography-alignRight css-1js305m-MuiTypography-root"
                >
                  Sat
                </h6>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1idn90j-MuiGrid-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-1bpfpyh"
        >
          <div
            class="css-e53awj-MuiStack-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-ns280z-MuiTypography-root"
            >
              Employees
            </h6>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-15n7dva-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              +
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="css-nen11g-MuiStack-root"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-scxkcp"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-iq9sbh-MuiTypography-root"
          >
            Tools
          </h6>
          <div
            class="css-4maiob-MuiStack-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Add Shift
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Export
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
      >
        <div
          class="css-1p5q5e5-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r0:"
              id=":r0:-label"
            >
              Schedule Name
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                type="text"
                value="12/29/2019, 12:00:00 AM"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Schedule Name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r1:"
              id=":r1:-label"
            >
              Week
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                placeholder="mm/dd/yyyy"
                type="tel"
                value="12/29/2019"
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <button
                  aria-label="Choose date, selected date is Dec 29, 2019"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CalendarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Week
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-j46zze-MuiPaper-root"
          >
            Total hours: 
            0.00
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
